
# summary of urchin selection experiment analysis

- Experiment run by April. Not sure when.
- all data is capture seq, exons + promoter regions. 
- 25 indivs were spawned. 10 females, 15 males? maybe
- low and ambient pH conditions
- genotyping at day 1 and day 7 in both conditions. Looking for shifts in AF.


## data processing

Reran all data processing:

- maf 0.025
- qual 75
- bi-allelic only
- avgdepth > 50 & avgdepth < 264
- 42,154 variants remain

## detecting changes in allele freq

- initially used cmh, D1 to D7, in control and treatment. 
- but, this might not be best approach
    + see: http://onlinelibrary.wiley.com/doi/10.1111/2041-210X.12810/full
- Argue that GLM approach probably more powerful
    + previous simulation studies didn't look at these methods

### GLM approach

using the model: glm(cbind(count1, count2) ~ pH+day+pH:day, family="quasibinomial", data=dat_in) 

The quasibinomial faminily controls for overdispersion

after FDR correction (qvalue), no significant effects of either pH or pH*day interaction

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/glm_ph_hist.png" width="500">

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/glm_interaction_hist.png" width="500">

### CMH approach

Comparisons:

1. D1 control vs D7 low pH
2. D1 vs D7 control
3. D7 low pH vs D7 control

Significance defined as: max p value of the top 0.01 quantile of D1 vs D7 control. 

- D1 control vs D7 low pH
	+ 818 significant
- D1 vs D7 control: 
	+ ~400 sig (ie, 40k * 0.01)
- D7 low pH vs D7 control
	+ only 97 significant

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/pval_hist.png" width="500">

#### Patterns of LD

calculated using LDx. 

- Calcs LD based on within read LD. so only up to ~600 bp. 
- these data get noisy in our selected loci at greater distances. around 200 bp or so. 

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/ld_decay_4.png" width="500">

It looks like the D7 comparisons have minimal elevated LD

However, the selected loci are a subset of the genome. We can subsample the genome to see how these compare.

Following are 500 subsampled distributions of LD. The left is D1 vs D7, right is D7 comparison

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/ld_1.png" width="700">

looking at results of KS test for each of these permutations vs the raw data:

using both raw data and the model fits

- D1 vs D7 comparisons are all highly significant for both raw and model
	+ model fits are all pval of ~0
- D7 comparisons:
	+ raw data: very few sig
	+ model fits: more, but only about .5 < 0.05. 

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/ld_ks.png" width="700">


Now histograms of the esimated coefficients and where our observed values fall. I'm not positive this is the best way to compare these, but its the best I can come up with right now. 

- a is the intercept. 
- b is the depth of the curve. where it asympotyes

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/ld_3.png" width="700">


##### look only at variants that overlap between the two analyses

only 55 variants overlap between the two approaches.

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/ld_2.png" width="500">

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/ld_4.png" width="500">

p=0.09 for a


## variation in allele frequency estimates within groups

Would expect that: 

1. variance in AF should be lowest in day 1 pH 8. 
2. because of drift (maybe) higher variance in day 7 pH 8
3. day 7 pH 7.5 lowest? because selection to contstrain AF

order of variance: day1 pH8 < day7 pH8 < day7 pH7

all are statistically significant

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/af_sd_1.png" width="500">

also wanted to compare the variance in selected AF in each group

randomly selected 818 variants from each line, calculated median AF, repeated 2000 times. vertical lines represent the median sd in allele frequency in selected variants in that group (those that were identified as under selection in the D1 to D7 comparison). 

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/af_sd_2.png" width="500">

## comparing AF change across time

<img src="https://github.com/rsbrennan/urchin_af/tree/master/figures/af_change_comp.png" width="500">


## Moving forward with D1 to D7 comparison

choosing D1pH8 vs D7pH7.5 because:  
- seems like increased variation in D7_pH8 response is limiting statistical power  
- I can't think of any real reason to choose that the D7 comparison is preferred  
	+ I acutually like the paired D1-D7 tests to we can show encrichment in significance in the low pH treatment  

### coding vs non-coding regions

for total probe set:

- exons: 33092 variants (79%), 8001 probes (75%)
- non-coding: 9062, 2587 probes

so only 10588 probes represented in the data set. I think there were 41k? 21k genes? not sure. 
    
For the 818 sig genes:

- exons: 655 (80%), 471 probes (79%)
- non-coding: 163 variants, 126 probes













